src/gt4py/next/ffront/ast_passes/single_static_assign.py

```diff
@@ -48,45 +113,101 @@ def foo():
         a__2 = 3 + a__1
         return a__2
 
-    Note that each variable name is assigned only once and never updated / overwritten.
+    Note that each variable name is assigned only once (per branch) and never updated / overwritten.
 
     Note also that after parsing, running the pass and unparsing we get invalid but
     readable python code. This is ok because this pass is not intended for
     python-to-python translation.
+
+    WARNING: This pass is not intended as a general-purpose SSA transformation.
+    The pass does not support any general Python AST. Known limitations include:
+        * Nested functions aren't supported
+        * While loops aren't supported
     """
 
-    class RhsRenamer(ast.NodeTransformer):
-        """
-        Rename right hand side names.
+    assignment_tracker: _AssignmentTracker = dataclasses.field(default_factory=_AssignmentTracker)
 
-        Only read from parent visitor state, should not modify.
-        """
+    @classmethod
+    def apply(cls, node: ASTNodeT) -> ASTNodeT:
+        return cls().visit(node)
+
+    def _rename(self, node: ASTNodeT) -> ASTNodeT:
+        for child_node in ast.walk(node):
+            if isinstance(child_node, ast.Name):
+                child_node.id = unique_name(
+                    child_node.id, self.assignment_tracker.count(child_node.id)
+                )
+        return node
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> ast.FunctionDef:
+        # For practical purposes, this is sufficient, but really not general at all.
+        # However, the algorithm was never intended to be general.
+
+        old_versioning = self.assignment_tracker.copy()
+
+        for arg in node.args.args:
+            self.assignment_tracker.define(arg.arg)
+
+        node.body = [self.visit(stmt) for stmt in node.body]
+
+        self.assignment_tracker = old_versioning
+        return node
 
-        @classmethod
-        def apply(cls, name_counter, separator, node):
-            return cls(name_counter, separator).visit(node)
+    def visit_If(self, node: ast.If) -> ast.If:
+        old_versioning = self.assignment_tracker
 
-        def __init__(self, name_counter, separator):
-            super().__init__()
-            self.name_counter: dict[str, int] = name_counter
-            self.separator: str = separator
+        node.test = self._rename(node.test)
 
-        def visit_Name(self, node: ast.Name) -> ast.Name:
-            if node.id in self.name_counter:
-                node.id = f"{node.id}{self.separator}{self.name_counter[node.id]}"
+        self.assignment_tracker = old_versioning.copy()
```

Comment:
Wouldn't it be simpler to remove line 159 and here just do this?
```suggestion
        old_versioning = self.assignment_tracker.copy()
```